### ISSUE TITLE
Type annotation fails when the field name matches type name

### ISSUE THREAD

**Issue by Harry-Lees** [microsoft/pyright#3924]
**Describe the bug**
If a field name matches the name of the type, the annotation will fail with `Type of "x" could not be determined because it refers to itself`

**Screenshots or Code**
Minimal example shown below:

```python
from dataclasses import dataclass
from datetime import datetime

@dataclass
class Foo:
    datetime: datetime
```

Output from pyright:

```
(.venv) 18:05:53 harrylees@MacbookAir.local Documents python -m pyright example.py
No configuration file found.
No pyproject.toml file found.
stubPath /Users/harrylees/Documents/typings is not a valid directory.
Assuming Python platform Darwin
Searching for source files
Found 1 source file
pyright 1.1.270
/Users/harrylees/Documents/example.py
  /Users/harrylees/Documents/example.py:7:15 - error: Type of "datetime" could not be determined because it refers to itself (reportGeneralTypeIssues)
  /Users/harrylees/Documents/example.py:7:15 - error: Illegal type annotation: variable not allowed unless it is a type alias (reportGeneralTypeIssues)
2 errors, 0 warnings, 0 informations
Completed in 0.494sec
```

**VS Code extension or command-line**
This seems to have been introduced in version `1.1.270`. I ran the above example in terminal with `python -m pyright example.py` using `Python 3.10.6`.

**Additional context**

MyPy shows no issues with the example:

```
(.venv) 18:06:28 harrylees@MacbookAir.local Documents python -m mypy example.py
Success: no issues found in 1 source file
```

Pyright version `1.1.269` shows no issues with the example

```
(.venv) 18:08:25 harrylees@MacbookAir.local Documents python -m pyright example.py
WARNING: there is a new pyright version available (v1.1.269 -> v1.1.270).
Please install the new version or set PYRIGHT_PYTHON_FORCE_VERSION to `latest`

No configuration file found.
No pyproject.toml file found.
stubPath /Users/harrylees/Documents/typings is not a valid directory.
Assuming Python platform Darwin
Searching for source files
Found 1 source file
pyright 1.1.269
0 errors, 0 warnings, 0 informations
Completed in 0.465sec
```

**Comment by vrslev** [comment]
@erictraut I encountered the same bug as well. Yes, it is uncommon when a class attribute is named exactly like its type declaration. Although, I think it has its place.

`foo: int = foo` is not the same as `foo: foo`.

In my project there're quite a lot of classes that are just "data", they describe API endpoint schemas. Take a look at this file: https://github.com/vrslev/ikea-api-client/blob/59bfb505b9a75ca5e3fafff86539f82720763e04/src/ikea_api/wrappers/parsers/iows_items.py. I think is this case there's no more intuitive way to name classes that describe a JSON values of given keys.

Also, an example of common use case that would now fail:

```python
class C:
    type: type
```

**Comment by erictraut** [comment]
> I think it has its place.

You're entitled to your own opinion, but as I mentioned above, I would recommend against it because it leads to code fragility. It's like setting landmines in your code that will potentially blow up at some point in the future — affecting either you or whoever else is maintaining your code at the time. If someone on my team attempted to commit code like this, I'd reject the PR.

**Comment by vrslev** [comment]
To let me understand your point, can you tell me, what this snippet should look like then?

```python
class C:
    type: type
```

**Comment by erictraut** [comment]
The goal here is to avoid a namespace conflict. The fragility comes from the fact that the scoping rules and resolution order for type annotations are inconsistent (e.g. differ based on whether it's in a class body or a function body and whether it's in a type stub file versus a "py" file), are interpreted differently by different tools, and are likely to change in the future.

First, I'd try to avoid using a field name of `type` because it's already a built-in class name and it's not a very descriptive name. (Also, I'll point out that `type`, when used as a type annotation, is a generic that takes one type argument. If you can avoid an `Any`, that's preferable.)

```python
class C:
    factory: type[FactoryBase]
```

Second, if the conflicting field name `type` is unavoidable, then you have a few options to avoid the namespace conflict:

1. Provide an unambiguous local alias for `builtins.type`:
```python
from builtins import type as type_

class C:
    type: type_[Any]
```

2. Define a local type alias:
```python
FactoryType = type[FactoryBase]

class C:
    type: FactoryType
```

**Comment by vrslev** [comment]
That's helpful, thanks. Still, I think it is more of a principal rather than a good practice I'm talking about.

`Type of "..." could not be determined because it refers to itself` doesn't describe the problem. It talks about `foo: int = foo`, not `foo: foo`. I guess fixing the `foo: int = foo` bug introduced another bug, where `foo: foo` is interpreted as `foo = foo`. Do I understand this right? If yes, then what we're talking about here? _That's a bug._

**Comment by vrslev** [comment]
@Harry-Lees 

> I did notice that you were using Pydantic in the example you gave, I'm sure there must be a way around that in Pydantic, could you supply an alias to load data from the JSON under a different name?

Yes, you can do that. It is not always practical though. In my case I have quite a few API endpoint with a lot of "throw-away" information. And receiving and parsing the data using Pydantic is not enough, I have to completely remodel it to achieve expected result.

**Comment by erictraut** [comment]
This will be addressed in the next release.

**Comment by vrslev** [comment]
Thanks! 🙏

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.271, which I just published. It will also be included in a future release of pylance.
